File: CampaignSystem/Entities/Campaign.cs
using CampaignSystem.Enums;

namespace CampaignSystem.Entities;

/// <summary>
/// Campaign definition. Maps to CAMPAIGN.
/// The scope of a campaign is never hardcoded — it is read from the four criteria
/// junction tables. An empty junction table means the campaign is unrestricted on
/// that dimension.
/// </summary>
public class Campaign
{
    public int Id { get; set; }

    public string Name { get; set; } = null!;

    public string? Description { get; set; }

    public CampaignType CampaignType { get; set; }

    /// <summary>
    /// Only meaningful when <see cref="CampaignType"/> is <see cref="Enums.CampaignType.EnrollmentRequired"/>
    /// — null for a MASS campaign, which has no enrollment to measure from.
    /// </summary>
    public EnrollmentBasis? EnrollmentBasis { get; set; }

    public DateTime StartDate { get; set; }

    public DateTime EndDate { get; set; }

    /// <summary>Lower bound a single transaction must reach to qualify.</summary>
    public decimal? MinimumAmount { get; set; }

    /// <summary>Upper bound a single transaction must stay under to qualify.</summary>
    public decimal? MaximumAmount { get; set; }

    /// <summary>Points earned per qualifying transaction.</summary>
    public decimal? RewardPoint { get; set; }

    /// <summary>
    /// Reward cap. Applied to each reward row, so it follows whatever unit the campaign
    /// accumulates in: per customer under <see cref="EarningType.CustomerBased"/>, per card
    /// under <see cref="EarningType.CardBased"/>.
    /// </summary>
    public decimal? MaxRewardAmount { get; set; }

    /// <summary>
    /// Whether refunding a counted purchase claws its points back after the campaign has paid.
    /// Off means a refund is ignored once the reward is granted.
    /// </summary>
    public bool RefundClawbackEnabled { get; set; }

    /// <summary>
    /// How many days after the reward is loaded a refund can still claw points back. Null while
    /// clawback is enabled means no time limit; ignored when clawback is off.
    /// </summary>
    public int? RefundClawbackDays { get; set; }

    public EarningType EarningType { get; set; }

    /// <summary>
    /// Restricts the campaign to one gender. Null means no restriction, which is the same
    /// rule the criteria junction tables follow: saying nothing includes everyone.
    /// </summary>
    public Gender? Gender { get; set; }

    /// <summary>
    /// Restricts the campaign to primary or to supplementary cards. Null covers both — the
    /// "Hepsi" option on the screen this replaces.
    /// </summary>
    public CardType? CardType { get; set; }

    /// <summary>
    /// Where the campaign stands: Pending, Ongoing, Loading or Ended.
    ///
    /// Stored rather than worked out from the dates on every read, so that a query, a report
    /// or anyone looking at the table sees the same answer the application does. Keeping it
    /// truthful is the daily batch's job — it advances campaigns whose start or end date has
    /// arrived, then loads the rewards of the ones that are due.
    /// </summary>
    public CampaignStatus Status { get; set; }

    /// <summary>
    /// Whether transactions accumulate per card, which decides how many reward rows one
    /// customer receives. Kept here rather than in a service so the reward batch and the
    /// enrollment rules cannot read it differently.
    /// </summary>
    public bool AccumulatesPerCard => EarningType == EarningType.CardBased;

    public bool IsActive { get; set; } = true;

    public ICollection<CampaignSegment> CampaignSegments { get; set; } = [];

    public ICollection<CampaignProduct> CampaignProducts { get; set; } = [];

    public ICollection<CampaignMerchant> CampaignMerchants { get; set; } = [];

    public ICollection<CampaignTransactionCode> CampaignTransactionCodes { get; set; } = [];

    public ICollection<CampaignCondition> Conditions { get; set; } = [];

    public ICollection<CampaignParticipation> Participations { get; set; } = [];

    public ICollection<CampaignReward> Rewards { get; set; } = [];
}

File: frontend/src/app/models/campaign.ts
/**
 * Mirrors CampaignDto on the server. The property names have to match the JSON exactly,
 * so they are camelCase here — ASP.NET Core serialises that way by default.
 *
 * The enums arrive as strings rather than numbers because the API registers
 * JsonStringEnumConverter, which is also why these are string unions and not numeric enums:
 * "Mass" is what actually comes down the wire.
 */
export type CampaignType = 'Mass' | 'EnrollmentRequired';

export type EarningType = 'CardBased' | 'CustomerBased';

export type CampaignStatus = 'Pending' | 'Ongoing' | 'Loading' | 'Ended';

export type Gender = 'Male' | 'Female';

export type CardType = 'Primary' | 'Supplementary';

export interface Campaign {
  id: number;
  name: string;
  description: string | null;
  campaignType: CampaignType;
  startDate: string;
  endDate: string;
  minimumAmount: number | null;
  maximumAmount: number | null;
  rewardPoint: number | null;
  maxRewardAmount: number | null;
  earningType: EarningType;

  /** Null when the campaign is open to every gender. */
  gender: Gender | null;

  /** Null when the campaign covers both primary and supplementary cards. */
  cardType: CardType | null;

  status: CampaignStatus;
  isActive: boolean;
}

/**
 * What the create endpoint accepts. Id, status and isActive are absent because the server
 * decides them: a new campaign is always an active draft whose status follows from its dates.
 */
export interface CreateCampaign {
  name: string;
  description: string | null;
  campaignType: CampaignType;
  earningType: EarningType;

  // Null rather than omitted: the API reads null as "no restriction", which is what an
  // empty choice on the form means.
  gender: Gender | null;
  cardType: CardType | null;

  startDate: string;
  endDate: string;
  minimumAmount: number | null;
  maximumAmount: number | null;
  rewardPoint: number | null;
  maxRewardAmount: number | null;
}

/**
 * A campaign's scope. An empty list means the campaign places no restriction on that
 * dimension — not that it matches nothing.
 */
export interface CampaignCriteria {
  segmentIds: number[];
  productIds: number[];
  merchantIds: number[];
  transactionCodeIds: number[];
}
